### PR TITLE
fix(gpu): make training actually update weights on the CUDA path

### PR DESCRIPTION
**Pull request by bararchy**
## Symptom

Training a feedforward net (`135 -> 64 -> 32 -> 15`, all sigmoid, multi-label) with `-Denable_cuda` produced a completely flat error; the same code converged on CPU.

```
Epoch: 0,   Error: -0.86174,  MSE: 0.049506
Epoch: 200, Error: -0.862413, MSE: 0.049584
Epoch: 400, Error: -0.862413, MSE: 0.049584   <- weights never change
```

## Root causes (four, all GPU-only)

**1. CPU-computed output gradients never reached the device.**
`compute_cost_and_gradient_cpu` writes the cost derivatives with `unsafe_set`, which by design does *not* set `@host_modified`. The following `sync_to_device!` therefore hit its `return unless @host_modified` early return and copied nothing, while `mark_device_dirty!` was called regardless — so `backward` consumed the still-zeroed device buffer. Every gradient was zero.

**2. Adam updates never reached the device.**
`update_weights_adam_gpu` syncs down, computes the update on the host with `unsafe_set`, then syncs back — the same silent no-op. The new weights stayed on the host and were discarded by the next `sync_from_device!`.

**3. The batch workspaces were donated to the shared pool while still in use.**
`process_batch` returned `@batch_in_ws` / `@batch_out_ws` to the workspace pool after every sample, even though the network keeps and reuses them. `get_workspace` then handed the same buffer back out as a temporary — the output layer's forward scratch has the same shape as the expected-output workspace — zeroed it and overwrote it mid-sample. This corrupted the target and made the error *diverge* once (1) and (2) were fixed.

**4. CUDNN softmax cross-entropy was used unconditionally.**
Every GPU batch went through `CUDNN.softmax_cross_entropy_loss_and_gradient`, ignoring the requested cost function and the output activation. With a sigmoid output layer the activation is applied twice (sigmoid, then CUDNN's internal softmax); for a single-output layer softmax is identically `1.0`, so the loss was exactly `0.0` and training stopped at epoch 0 on the error threshold. It is now only used when the output layer emits raw logits.

Also fixed, same defect class (`unsafe_set`/`raw_data` write followed by a skipped `sync_to_device!`): `softmax_rows!`, the cuDNN element-mul / dropout / tanh CPU fallbacks, the `run()` activation fallback, and `to_matrix`, which `memcpy`'d a `Slice(Float64)` into an F32 device buffer using `size * 4` bytes (garbage on the device) and then marked it clean, so the correct host values were never uploaded.

## Verification

Single-step probe with fixed weights, one sample — GPU now bit-matches CPU:

```
=== CPU ===                                  === GPU ===
forward: [0.609631]                          forward: [0.609631]
olayer.g_w: -0.049928, -0.04413, -0.056729   olayer.g_w: -0.049928, -0.04413, -0.056729
hidden.g_w: -0.016166, 0.018534, -0.019879   hidden.g_w: -0.016166, 0.018534, -0.019879
olayer.weights: 0.70025, -0.799779, 0.900284 olayer.weights: 0.70025, -0.799779, 0.900284
```

XOR with Adam, GPU (was flat at MSE 0.25):

```
Epoch: 0,    Error: 0.125107, MSE: 0.015652
Epoch: 1000, Error: 0.103422, MSE: 0.010696
Epoch: 1500, Error: 0.059784, MSE: 0.003574
Final outputs: [0.2234, 0.7408, 0.7646, 0.2708]   # targets 0,1,1,0
```

Reported architecture reproduced (135->64->32->15 sigmoid, 1277 samples, sparse multi-label, `mini_batch_size: 32`, Adam), 40 epochs on GPU: **MSE 0.253776 -> 0.022988 (-90.9%)**.

Full suite green both ways: `179 examples, 0 failures` with `-Denable_cuda` (5 pending) and without (26 pending). `crystal tool format --check` and `ameba` clean.

Adds `spec/training_convergence_spec.cr`, which seeds the global RNG (weight init draws from it) and asserts both that the error drops and that every output-layer and hidden-layer weight actually moved. It fails 2/2 examples on master with CUDA enabled and passes on both the CUDA and CPU paths here.

## Behaviour note

Fix (4) changes what the GPU path optimises when you ask for `:mse` with an activated output layer: previously it silently optimised softmax cross-entropy instead. `spec/integration_spec.cr` (iris) was leaning on that — softmax CE happens to be a good objective for 3-class classification, so it reached ~0.6 accuracy in 100 epochs, whereas real MSE lands right on the spec's loose `> 0.2` threshold and flakes. Bumped to 600 epochs, which reaches 0.86–0.98 in about a second.

Full GPU suite run 5x consecutively after the change: `179 examples, 0 failures` each time.

### CODE REVIEW

File: spec/training_convergence_spec.cr [hunk 1]
```diff
@@ -0,0 +1,88 @@
+require "./spec_helper"
+
+# Regression coverage for the GPU training path, which used to leave weights
+# completely unchanged (flat MSE) because CPU-computed gradients and Adam
+# updates were never uploaded to the device, and because the per-batch input /
+# expected workspaces were handed back to the shared workspace pool while still
+# in use.
+describe "training convergence" do
+  it "drives XOR error down with adam" do
```
[the remark this answers is not in the document]
Comment (bararchy) [reply spec/training_convergence_spec.cr:22]: Fair — `MatrixLayer#random_fill!` draws from the global RNG, so the convergence thresholds were not deterministic across runs. Seeded per example with `Random::DEFAULT.new_seed`, matching `spec/multi_head_attention_cuda_spec.cr`. Fixed in b9107d8.

File: spec/training_convergence_spec.cr [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      cost_function: :mse,
+      epochs: 3000,
+      error_threshold: -1.0,
+      mini_batch_size: data.size,
+      log_each: 10_000
+    )
+
+    after = mse.call
+    after.should be < before * 0.5
+  end
```
[the remark this answers is not in the document]
Comment (bararchy) [reply spec/training_convergence_spec.cr:60]: Correct, and the mismatch was mine: the weights-changed assertion was in the first version of the spec, then I swapped it for the error-drop check because on its own it still passed on master. Both are in now — every output-layer and hidden-layer weight must have moved, *and* the error must drop. Fixed in b9107d8. Verified the spec still fails on master under CUDA (2/2 examples, on the error assertions) and passes on this branch on both the CUDA and CPU paths.
